### PR TITLE
Automatic Exposure and Gain Control

### PR DESCRIPTION
**Pull request by jpratt003**
Multi-tier exposure/gain control algorithm. Each band is configured as a range, where either exposure or gain is locked in that range and the other is controlling the output intensity.
Range 0: Gain: 0dB (locked) Exposure: <20ms
Range 1: Gain: <15dB. Exposure 20ms (locked)
Range 2: Gain: 15dB (locked). Exposure: <33ms
Range4: Gain: <40dB. Exposure: 33ms (locked).

### CODE REVIEW

File: launch/blackfly_s_autoexposure.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+def generate_launch_description():
+    """Generate launch description with multiple components."""
+    flir_dir = get_package_share_directory('flir_spinnaker_ros2')
+    config_dir = flir_dir + '/config/'
+    name_arg = LaunchArg('camera_name', default_value='blackfly_s',
+                         description='camera name')
+    serial_arg = LaunchArg('serial', default_value="'20435008'",
+                         description='serial number')
+    print([LaunchConfig('serial'),'_'])
```
Comment (josh-wende) [inline launch/blackfly_s_autoexposure.py:41]: this print statement still needed?

File: src/camera_autoexposure.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+{
+
+  declare_parameter("camera_name", "blackfly_s");
+  get_parameter("camera_name", camera_name_);
+  declare_parameter("target_intensity", 120);
+  get_parameter("target_intensity", target_intensity_);
+  declare_parameter("intensity_update_gain", 0.5);
+  get_parameter("intensity_update_gain", intensity_update_gain_);
+  // image_sub_ = create_subscription<sensor_msgs::msg::Image>(
+  //     "/" + camera_name_ + "/image_raw", 1, std::bind(&CameraAutoExposure::imageCallback, this, std::placeholders::_1));
```
Comment (josh-wende) [inline src/camera_autoexposure.cpp]: if this won't be needed later, should probably be deleted.

File: src/camera_autoexposure.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+CameraAutoExposure::~CameraAutoExposure() {
+
+}
+
+void CameraAutoExposure::metaCallback(
+  const image_meta_msgs_ros2::msg::ImageMetaData::ConstSharedPtr msg) {
+  if (msg->brightness == -1) {
+    RCLCPP_WARN(get_logger(), "Brightness is disabled, not running exposure control.");
+  }
```
Comment (josh-wende) [inline src/camera_autoexposure.cpp]: in this case, shouldn't it prevent the rest of the function's code from running?

File: src/camera_autoexposure.cpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  } else if (target_combined_exposure_gain < (max_exposure_time * std::pow(10, kMaxGain / 20.0))) {
+    camera_control_msg.exposure_time = max_exposure_time;
+    const double target_linear_gain = target_combined_exposure_gain / max_exposure_time;
+    camera_control_msg.gain = 20 * std::log10(target_linear_gain);
+  } else {
+    camera_control_msg.gain = kMaxGain;
+    camera_control_msg.exposure_time = max_exposure_time;
+  }
+
+  static int skip_count = 20;
```
Comment (josh-wende) [inline src/camera_autoexposure.cpp]: I've been told this is not good practice--consider refactoring.

File: launch/blackfly_s_autoexposure.py (hunk 1 above)
Comment (jpratt003) [reply launch/blackfly_s_autoexposure.py:41]: It's a useful reminder when you launch. If you don't update the serial no, it's generally wrong.

**Pull request merged**